### PR TITLE
feat: Add time value type

### LINKED ISSUE #1397: Missing definition "TimeValue" in knora-base
Thereis no "TimeValue" in knora-base in order to store a timestamp as "normal" value of a resource.

Needs https://github.com/dhlab-basel/Knora-ui/issues/307.
Needs https://github.com/dhlab-basel/knora-py/issues/12.
Needs https://github.com/dhlab-basel/knora-api-js-lib/issues/41.

### LINKED ISSUE #1446: Compiler warnings
```bash
[info] Compiling 402 Scala sources to .../webapi/target/scala-2.12/classes ...
[warn] .../webapi/src/main/scala/org/knora/webapi/store/triplestore/embedded/JenaTDBActor.scala:441:49: class RAMDirectory in package store is deprecated: see corresponding Javadoc for more information.
[warn]             val indexDirectory: Directory = new RAMDirectory()
[warn]                                                 ^
[warn] .../webapi/src/main/scala/org/knora/webapi/store/triplestore/embedded/JenaTDBActor.scala:452:49: class RAMDirectory in package store is deprecated: see corresponding Javadoc for more information.
[warn]             val indexDirectory: Directory = new RAMDirectory()
[warn]                                                 ^
[warn] two warnings found

```

### PR DESCRIPTION
**Pull request by benjamingeer**
This adds a new value type, `knora-base:TimeValue`, which stores an `xsd:dateTime` (represented as `xsd:dateTimeStamp` in API v2).

- [x] Add `TimeValue` to `knora-base` and `salsah-gui`.
- [x] Support `knora-base:TimeValue` in SPARQL templates:
  - [x] in API v1
  - [x] in API v2
- [x] Support time values in API v1:
  - [x] reading
  - [x] writing
  - [x] bulk import
  - [x] extended search
- [x] Support time values in API v2:
  - [x] reading
  - [x] writing
  - [x] Gravsearch
- [x] Update tests.
  - [x] `ValuesResponderV1Spec`
  - [x] `ValuesV1R2RSpec`
  - [x] `ResourcesV1R2RSpec`
  - [x] `ValuesResponderV2Spec`
  - [x] `ValuesRouteV2E2ESpec`
  - [x] `OntologyV2R2RSpec`
    - [x] Update RDF4J library (fixes stack overflow error).
  - [x] `OntologyResponderV2Spec`
  - [x] `SearchV1R2RSpec`
  - [x] `SearchRouteV2R2RSpec`
  - [x] test data returned for generated client API code
- [x] Add update script plugin.
- [x] Update docs.

Resolves #1397.
Resolves #1446.
Needs https://github.com/dhlab-basel/Knora-ui/issues/307.
Needs https://github.com/dhlab-basel/knora-py/issues/12.
Needs https://github.com/dhlab-basel/knora-api-js-lib/issues/41.

### CODE REVIEW

**Comment by tobiasschweizer** [comment]
@benjamingeer I will fix the anything ontology test data on this branch too.

**Comment by benjamingeer** [comment]
TODO: when this is ready to merge, make a new PR for it with an updated Knora version number.

**Comment by lrosenth** [comment]
@tobiasschweizer Könntest Du diesen Pullrequest anschauen und eventuell "approven"?

**Comment by tobiasschweizer** [comment]
I will check this PR next week.

**Comment by gfoo** [comment]
if you want to review this PR next week, do you think that we can plan to use it in our ontologies for the next Knora release?

**Comment by tobiasschweizer** [comment]
> if you want to review this PR next week, do you think that we can plan to use it in our ontologies for the next Knora release?

@subotic When should this be included?

**Comment by gfoo** [comment]
@tobiasschweizer  @subotic 

Made many tests with the branch, not found any bugs 👍 

I'm not yet sure to use this new type anyway. So for me, right now, no high priority for this pull request.

**Comment by gfoo** [comment]
> I'm not yet sure to use this new type anyway. So for me, right now, no high priority for this pull request.

Finally, interested by this new type, I'm using it in our ontology, until now API works fine. Would you plan this feature for the next release?

**Comment by benjamingeer** [comment]
@gfoo waiting for the issues marked "Needs" in the PR description

**Comment by tobiasschweizer** [comment]
Could generate test data containing a time value? So far, I have used https://github.com/dasch-swiss/knora-api-js-lib/test/data/api/v2/resources/thing-with-time-value.json

**Comment by tobiasschweizer** [comment]
I got it from here: /webapi/src/test/resources/test-data/searchR2RV2/ThingWithTimeStamp.jsonld

**Comment by benjamingeer** [comment]
@tobiasschweizer Is this OK to merge now?

**Comment by tobiasschweizer** [comment]
I can look at tit tomorrow when I am in Basel. I guess I will have to regenerate the test data to be sure it works. @subotic is working on automating this.

**Comment by tobiasschweizer** [comment]
@benjamingeer Could generate test data containing a time value? So far, I have used https://github.com/dasch-swiss/knora-api-js-lib/test/data/api/v2/resources/thing-with-time-value.json

**Comment by benjamingeer** [comment]
> Could generate test data containing a time value?

- `test-data/v2/resources/testding.json`
- `test-data/v2/values/get-time-value-response.json`
- `test-data/v2/values/create-time-value-request.json`
- `test-data/v2/values/update-time-value-request.json`

**Comment by tobiasschweizer** [comment]
thanks, I will integrate that tomorrow!

**Comment by tobiasschweizer** [comment]
Could you add a time value to `create-resource-with-values-request.json` so I can also test resource creation with time values?

**Comment by benjamingeer** [comment]
> Could you add a time value to create-resource-with-values-request.json so I can also test resource creation with time values?

Done.

**Comment by tobiasschweizer** [comment]
Thanks, I will it integrate this today!

However, there is a failing E2E test (possibly unrelated to what we are doing here).

**Comment by benjamingeer** [comment]
> there is a failing E2E test

Random out-of-memory error on GitHub...

**Comment by tobiasschweizer** [comment]
> Random out-of-memory error on GitHub...

That sounds consistent, easy to anticipate and straight forward to fix ...

File: webapi/src/test/scala/org/knora/webapi/e2e/v2/SearchRouteV2R2RSpec.scala [hunk 1]
```diff
@@ -217,18 +217,17 @@ class SearchRouteV2R2RSpec extends R2RSpec {
                   |    ?thing anything:hasInteger ?intVal .
                   |    anything:hasInteger knora-api:objectType xsd:integer .
                   |    ?intVal a xsd:integer .
-                  |    FILTER(?intVal = 123454321)
+                  |    FILTER(?intVal = 123454321 || ?intVal = 999999999)
```
Comment (tobiasschweizer) [inline webapi/src/test/scala/org/knora/webapi/e2e/v2/SearchRouteV2R2RSpec.scala]: Why did to extend the expression with `|| ?intVal = 999999999`?

**Comment by benjamingeer** [comment]
> Why did to extend the expression with || ?intVal = 999999999?

Adding more test data broke existing tests, which do things like "find a Thing resource that may have an an optional boolean", which would find any Thing resource. So when I added new test data, I gave each new resource an extra integer (in this case 999999999) so I could exclude it from the existing tests, so they would still pass.

**Comment by tobiasschweizer** [comment]
Have you added a test for a conversion of XML with a time value attribute to standoff and back?

**Comment by benjamingeer** [comment]
> Have you added a test for a conversion of XML with a time value attribute to standoff and back?

Nope. I'd need to make a mapping for that, right?

**Comment by tobiasschweizer** [comment]
> Nope. I'd need to make a mapping for that, right?

I guess you can also extend an existing mapping.

**Comment by benjamingeer** [comment]
> I guess you can also extend an existing mapping.

Is there an existing mapping other than the standard mapping? Or should I add it to the standard mapping?

**Comment by tobiasschweizer** [comment]
The standard mapping is used by the old SALSAH GUI (with CKeditor). But adding a new element to the mapping should be fine (even if it's not used by Ckeditor's configuration).

**Comment by benjamingeer** [comment]
To add something to the standard mapping, would I have to regenerate the whole mapping from an XML file?

**Comment by benjamingeer** [comment]
Easier to add it to the HTML mapping...

**Comment by tobiasschweizer** [comment]
> Easier to add it to the HTML mapping...

I guess so because their is no in-triplestore-representation that you would have to update

**Comment by benjamingeer** [comment]
I've tried to add a timestamp mapping element in https://github.com/dasch-swiss/knora-api/pull/1403/commits/1961a4d85e33b5ce938e7fe0b92b8773376ad24d, but I get an error when creating the mapping:

```
org.knora.webapi.BadRequestException:
the provided mapping could not be handled correctly:
the following required standoff properties are not defined for the standoff class
http://www.knora.org/ontology/knora-base#StandoffTimeTag:
http://www.knora.org/ontology/knora-base#valueHasTimeStamp
```

What have I done wrong? You can run the test like this:

```
webapi/testOnly *SearchRouteV2R2RSpec -- -z t1
```

**Comment by tobiasschweizer** [comment]
Does `StandoffTimeTag ` have cardinalities for `valueHasTimeStamp`? Could you compare it to another existing standoff class?

**Comment by benjamingeer** [comment]
> Does StandoffTimeTag have cardinalities for valueHasTimeStamp? Could you compare it to another existing standoff class?

It does, and I did.

```ttl
:StandoffTimeTag rdf:type owl:Class ;

                        rdfs:subClassOf :StandoffDataTypeTag,
                                        :TimeBase ;

                        rdfs:comment "Represents a timestamp in a TextValue"@en .


:TimeBase rdf:type owl:Class ;

              rdfs:subClassOf :ValueBase ,
                              [ rdf:type owl:Restriction ;
                                owl:onProperty :valueHasTimeStamp ;
                                owl:cardinality "1"^^xsd:nonNegativeInteger
                              ] .


:valueHasTimeStamp rdf:type owl:DatatypeProperty ;

                       rdfs:subPropertyOf :valueHas ;

                       :subjectClassConstraint :TimeBase ;

                       :objectDatatypeConstraint xsd:dateTime .
```

**Comment by tobiasschweizer** [comment]
I can try and check this in the afternoon if you want

**Comment by benjamingeer** [comment]
Yes please. I based the mapping element on the existing one for the date tag, and I can't find anything that looks different.

**Comment by benjamingeer** [comment]
Found the bug...

**Comment by benjamingeer** [comment]
Fixed more bugs by adding the test with a standoff timestamp tag (https://github.com/dasch-swiss/knora-api/pull/1403/commits/f70fe6c90ec2931dab7a0c63d1234fc62a3bc1d7).

**Comment by tobiasschweizer** [comment]
With which release should this be integrated?

**Comment by benjamingeer** [comment]
> With which release should this be integrated?

It's not going into today's release. It could go into next month's release. Let's talk about this process at tomorrow's meeting.

**Comment by tobiasschweizer** [comment]
Ok! I would like to assign related PRs in knra-api-js-lib the same milestone tags. This should make it easy to keep track.

**Comment by benjamingeer** [comment]
> Can the SALSAH GUI display a time value? Are the value type constants correctly configured?

You mean the old GUI? I guess it can't. Which constants are those?

**Comment by tobiasschweizer** [comment]
> Which constants are those?

I think they are hard-coded in a JS file. They had to be changed to the Knora IRIs. Since the time value hasn't been supported yet by Knora, it is possible that the value type IRI has to be adapted. Maybe @loicjaouen knows this.

**Comment by benjamingeer** [comment]
OK, I think I've got it working in SALSAH 1 in 68ab4d5. I didn't really know what I was doing, but I can create a resource with a time value, display it, and change the time value. Could you have a look?

**Comment by tobiasschweizer** [comment]
> Could you have a look?

I haven't run it in years. @loicjaouen Could you have a look?

**Comment by benjamingeer** [comment]
@subotic I'm having trouble finding anyone who can review the changes I made in the SALSAH 1 JavaScript code for this PR. Could you help find a reviewer, or should we just merge it anyway?

**Comment by loicjaouen** [comment]
sorry, I missed the call, I look at it now

**Comment by benjamingeer** [comment]
Thanks @loicjaouen !

**Comment by benjamingeer** [comment]
@tobiasschweizer Is this OK to merge now?

**Comment by tobiasschweizer** [comment]
I will look at it in a minute!

**Comment by benjamingeer** [comment]
@tobiasschweizer Thank you!

**Pull request merged**